File: week-1/problem-set-recursion/prob6.py
"""Prob6 example.

This file belongs to the week 1 recursion practice solutions. It checks whether
strings are palindromes by comparing the first and last characters recursively.
"""


def clean_text(text):
    """Remove punctuation and spaces so the palindrome test is easier to read.

    Args:
        text (str): The text input used by this function.

    Returns:
        str: The string produced by the algorithm.
    """
    return "".join(character.lower() for character in text if character.isalnum())


def is_palindrome(s, length):
    """Return True when the given text reads the same forwards and backwards.

    Args:
        s (str): The s input used by this function.
        length (int): The length input used by this function.

    Returns:
        bool: True when the required condition is satisfied, otherwise False.
    """
    if length <= 1:
        return True

    if s[0] != s[length - 1]:
        return False

    return is_palindrome(s[1 : length - 1], length - 2)


def main():
    """Run the small demonstration for this teaching example.

    Returns:
        None: This function updates data or prints results and does not return a value.
    """
    examples = [
        "racecar",
        "hello",
        "A man, a plan, a canal, Panama!",
    ]

    for text in examples:
        cleaned = clean_text(text)
        print(f"{text!r} -> {is_palindrome(cleaned, len(cleaned))}")


if __name__ == "__main__":
    main()

File: week-1/problem-set-recursion/prob6.cpp
/**
 * @file prob6.cpp
 * @brief C++ version of the recursive palindrome example.
 */

#include <cctype>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

/**
 * @brief Remove spaces and punctuation so the palindrome test is easier to read.
 * @param text const string& Original text that may contain spaces or punctuation.
 * @return string Lowercase string containing only alphanumeric characters.
 */
string clean_text(const string& text) {
    string cleaned;
    for (char character : text) {
        if (isalnum(static_cast<unsigned char>(character))) {
            cleaned.push_back(static_cast<char>(tolower(static_cast<unsigned char>(character))));
        }
    }
    return cleaned;
}

/**
 * @brief Return true when the string reads the same forwards and backwards.
 * @param text const string& Cleaned text used for the recursive comparison.
 * @param length int Number of characters currently considered by the recursive call.
 * @return bool True when text is a palindrome, otherwise false.
 */
bool is_palindrome(const string& text, int length) {
    if (length <= 1) {
        return true;
    }
    if (text[0] != text[length - 1]) {
        return false;
    }
    return is_palindrome(text.substr(1, length - 2), length - 2);
}

/**
 * @brief Run the palindrome demonstrations.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<string> examples = {
        "racecar",
        "hello",
        "A man, a plan, a canal, Panama!"
    };

    for (const string& text : examples) {
        string cleaned = clean_text(text);
        cout << text << " -> " << boolalpha << is_palindrome(cleaned, static_cast<int>(cleaned.size())) << '\n';
    }
    return 0;
}
